### PR TITLE
Mutation routes

### PR DESCRIPTION
**Pull request by ckieverentineas**

### CODE REVIEW

File: schema/testing.js [hunk 1]
```diff
@@ -22,15 +22,74 @@ module.exports = (fastify, opts, next) => {
...
+
+  fastify.post('/categories/add', async ({ body, session }) => {
+    const [categories] = await Category.query().where({ name: body.name });
+    assert(!categories, new APIError(400, 'Not added test'));
+    await Category.query().insert({ name: body.name });
+    return 'OK';
+  });
+  fastify.post('/questions/add', async ({ body, session }) => {
+    const [questions] = await Question.query().where({ text: body.text });
+    assert(!questions, new APIError(400, 'Not added question'));
```
Comment (MurkyMeow) [inline schema/testing.js]: I think having questions which texts are the same is ok?

File: schema/testing.js [hunk 2]
```diff
@@ -22,15 +22,74 @@ module.exports = (fastify, opts, next) => {
...
     );
-    const rows = query.answer_ids.map(id => ({ user_id: session.userid, answer_id: id }));
+    const rows = body.answer_ids.map(id => ({ user_id: session.userid, answer_id: id }));
     await Progress.query().insertGraph(rows);
     return { answer: 'Ok' };
   });
+
+  fastify.post('/categories/add', async ({ body, session }) => {
+    const [categories] = await Category.query().where({ name: body.name });
+    assert(!categories, new APIError(400, 'Not added test'));
```
Comment (MurkyMeow) [inline schema/testing.js]:
```suggestion
    assert(!categories, new APIError(409, 'This name is busy'));
```

Comment (MurkyMeow) [reply schema/testing.js]: Also, why is it called `tests` if it only describes an item, not the list of items? The same appeals to the lines below.

File: schema/testing.js [hunk 3]
```diff
@@ -22,15 +22,74 @@ module.exports = (fastify, opts, next) => {
...
+  fastify.post('/tests/add', async ({ body, session }) => {
+    const [tests] = await Test.query().where({ name: body.name });
+    assert(!tests, new APIError(400, 'Not added test'));
+    await Test.query().insert({ name: body.name, category_id: body.category_id });
+    return 'OK';
+  });
+
+  fastify.post('/categories/edit', async ({ body, session }) => {
+    const [categories] = await Category.query().where({ id: body.id });
+    assert(categories, new APIError(400, 'Not edit test'));
```
Comment (MurkyMeow) [inline schema/testing.js]:
```suggestion
    assert(categories, new APIError(400, 'Requested category does not exist'));
```

Comment (MurkyMeow) [reply schema/testing.js]: Let's return the updated category instead of `Ok`. (Same for the routes below).

**Comment by MurkyMeow** [comment]
Please don't make commits if you have lint errors. Those commits are marked with the red cross, by pressing it you cant find what went wrong in your code.

File: schema/testing.js (hunk 2 above)
Comment (ckieverentineas) [reply schema/testing.js]: Maybe "This category already exists"?

**Comment by MurkyMeow** [comment]
Manually resolved the conflicts. Could you ensure that everything is working correct?

**Comment by MurkyMeow** [comment]
I've reduced all of the boilerplate code presented here to a single function in 5aeb92fd4781aa467fb1a23d71e515ce3dd122d3. Closing.

**Pull request closed without merging**